### PR TITLE
feat: Accept offers with onERC721Received

### PR DESCRIPTION
**Pull request by fzavalia**

### CODE REVIEW

File: contracts/Rentals.sol [hunk 1]
```diff
@@ -313,15 +275,30 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
...
+    ) external override returns (bytes4) {
+        // When calling acceptListing or acceptOffer, this contract will transfer this asset from the holder to itself to initialize a rent.
+        // There is no need to do anything extra if that is the case.
+        if (_operator != address(this)) {
+            Offer memory offer = abi.decode(_data, (Offer));
+
+            require(_msgSender() == offer.contractAddress, "Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH");
+
+            // Set the lessor of the rental to track that the contract already has it.
+            rentals[offer.contractAddress][offer.tokenId].lessor = _from;
```
Comment (nachomazzara) [inline contracts/Rentals.sol:297]: I would respect the CEI pattern, moving this after all of the requirements inside `_acceptOffer`

File: contracts/Rentals.sol [hunk 2]
```diff
@@ -381,22 +382,30 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
...
+        // Transfer tokens
         if (_pricePerDay > 0) {
             _handleTokenTransfers(_lessor, _tenant, _pricePerDay, _rentalDays);
         }
 
+        IERC721Rentable asset = IERC721Rentable(_contractAddress);
+
+        // If the provided contract support the verifyFingerpint function, validate the provided fingerprint.
+        if (asset.supportsInterface(InterfaceId_VerifyFingerprint)) {
+            require(asset.verifyFingerprint(_tokenId, abi.encodePacked(_fingerprint)), "Rentals#_rent: INVALID_FINGERPRINT");
```
Comment (nachomazzara) [inline contracts/Rentals.sol:401]: we can do a static call and leave this in the checks part

File: test/Rentals.spec.ts [hunk 3]
```diff
@@ -1569,21 +1569,81 @@ describe('Rentals', () => {
...
+
+      expect(rental.lessor).to.equal(lessor.address)
+      expect(rental.tenant).to.equal(tenant.address)
+      expect(rental.endDate).to.equal(latestBlockTime + daysToSeconds(offerParams.rentalDays))
+    })
+
+    it('should revert when the caller is different from the contract address provided in the offer', async () => {
+      const bytes = ethers.utils.defaultAbiCoder.encode([offerEncodeType], [offerEncodeValue])
+      await expect(rentals.onERC721Received(extra.address, lessor.address, tokenId, bytes)).to.be.revertedWith(
+        'Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH'
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1645]: I would try to add more edge cases and therefore more tests. Also, add the ones for the acceptOffer.

- Try to test every revert reason inside `_acceptOffer`
- Try to revert if the user is safe-transferring the asset for a re-rent while the asset is inside the rent contract
- Try to revert if the token Id being sent is not the one expected.

File: contracts/interfaces/IERC721Rentable.sol [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.7;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+/// @notice Extended ERC721 interface with methods required by the Rentals contract.
+interface IERC721Rentable is IERC721 {
+    /// @dev Updates the operator of the asset.
```
[the remark this answers is not in the document]
Comment (fzavalia) [reply contracts/interfaces/IERC721Rentable.sol]: Updated the comment with your suggestion but with a few syntactical changes.

File: contracts/Rentals.sol [hunk 5]
```diff
@@ -313,15 +274,30 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
...
+            Offer memory offer = abi.decode(_data, (Offer));
+
+            // Check that the caller is the contract defined in the offer to ensure the function is being
+            // called through an ERC721.safeTransferFrom.
+            require(msg.sender == offer.contractAddress, "Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH");
+
+            // Prevents the holder from sending another token than the one defined in the offer.
+            require(offer.tokenId == _tokenId, "Rentals#onERC721Received: TOKEN_ID_MISMATCH");
+
+            _acceptOffer(offer, _from);
```
Comment (nachomazzara) [inline contracts/Rentals.sol:298]: I guess that this should be _operator instead of _from

Comment (fzavalia) [reply contracts/Rentals.sol:298]: fixed

Comment (fzavalia) [reply contracts/Rentals.sol:298]: _operator works fine when the holder of an asset sends it to the contract via `safeTransferFrom` because the operator is always the address of the caller of that function.

However, if the asset is being sent by another contract that was given allowance, _operator will be that contract instead of the holder.

If that was the case, that contract would be considered as the lessor, and would be the only one able to claim it back.

Using from makes it so that the holder of the token is set as lessor.

Comment (fzavalia) [reply contracts/Rentals.sol]: Using `_operator` but added a comment in the contract to keep in mind a possible outcome

File: contracts/Rentals.sol [hunk 6]
```diff
@@ -313,15 +274,33 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
...
-        require(_operator == address(this), "Rentals#onERC721Received: ONLY_ACCEPT_TRANSFERS_FROM_THIS_CONTRACT");
+        uint256 _tokenId,
+        bytes memory _data
+    ) external override returns (bytes4) {
+        if (_operator != address(this)) {
+            Offer memory offer = abi.decode(_data, (Offer));
+
+            // Check that the caller is the contract defined in the offer to ensure the function is being
+            // called through an ERC721.safeTransferFrom.
+            require(msg.sender == offer.contractAddress, "Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH");
```
Comment (nachomazzara) [inline contracts/Rentals.sol:296]: Maybe we can use one require here to reduce some gas

Comment (fzavalia) [reply contracts/Rentals.sol:300]: https://github.com/decentraland/rentals-contract/pull/20
